lib/launcher.rb

```diff
@@ -149,32 +177,85 @@ def write_pid
       retry
     end
 
-    def webrick_server(app, addresses, port)
+    def add_puma_server(app, addresses, port, conn_type)
+      # IMPORTANT:
+      # The following code takes only a single host.
+      # The current reason for it, is that "run" is blocking, and in order to
+      # add support for more hosts, additional threads requires to be created
+      address = addresses.first
+      address = '0.0.0.0' if address == '*'
+      if conn_type == :ssl
+        host = "ssl://#{address}/"
+        require 'cgi'
+        query_list = []
+        app[:SSLArgs].each_pair do |name, value|
+          query_list << "#{CGI::escape(name.to_s)}=#{CGI::escape(value)}"
+        end
+        host = "#{host}?#{query_list.join('&')}"
+      else
+        host = address
+      end
+      Rack::Handler::Puma.run(app[:app],
+                              Verbose: true,
+                              Port: port,
+                              Host: host
+                             )
+    end
+
+    def add_webrick_server(app, addresses, port)
       server = ::WEBrick::HTTPServer.new(app)
-      addresses.each {|a| server.listen(a, port)}
-      server.mount "/", Rack::Handler::WEBrick, app[:app]
+      addresses.each { |a| server.listen(a, port) }
+      server.mount '/', Rack::Handler::WEBrick, app[:app]
       server
     end
 
+    def add_threaded_server(server_name, conn_type, app, addresses, port)
+      case server_name
+      when :webrick
+        Thread.new do
+          @servers << add_webrick_server(app, addresses, port).start
+        end
+      when :puma
+        Thread.new do
+          add_puma_server(app, addresses, port, conn_type)
```

Comment:
If this isn't added to `@servers`, how are the signal handlers attached?

Reply:
Resolved.